magnum-zio/src/main/scala/com/augustnagro/magnum/magzio/Transactor.scala

```diff
@@ -0,0 +1,173 @@
+package com.augustnagro.magnum.magzio
+
+import com.augustnagro.magnum.{DbCon, DbTx, SqlException, SqlLogger}
+import zio.{IO, Semaphore, Task, Trace, UIO, URIO, ZIO}
+
+import java.sql.Connection
+import javax.sql.DataSource
+import scala.reflect.ClassTag
+import scala.util.control.{ControlThrowable, NonFatal}
+
+class Transactor private (
+    dataSource: DataSource,
+    sqlLogger: SqlLogger,
+    connectionConfig: Connection => Unit,
+    semaphore: Option[Semaphore]
+):
+
+  def withSqlLogger(sqlLogger: SqlLogger): Transactor =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def withConnectionConfig(connectionConfig: Connection => Unit): Transactor =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def connect[A](f: DbCon ?=> A)(using Trace): Task[A] =
+    val zio = ZIO.blocking(
+      ZIO.acquireReleaseWith(acquireConnection)(releaseConnection)(cn =>
+        ZIO.attempt {
+          connectionConfig(cn)
+          f(using DbCon(cn, sqlLogger))
+        }
+      )
+    )
+    semaphore.fold(zio)(_.withPermit(zio))
+
+  def transact[A](f: DbTx ?=> A)(using Trace): Task[A] =
+    val zio = ZIO.blocking(
+      ZIO.acquireReleaseWith(acquireConnection)(releaseConnection)(cn =>
+        ZIO.attempt {
+          connectionConfig(cn)
+          cn.setAutoCommit(false)
+          try
+            val res = f(using DbTx(cn, sqlLogger))
+            cn.commit()
+            res
+          catch
+            case NonFatal(t) =>
+              cn.rollback()
+              throw t
+        }.uninterruptible
+      )
+    )
+    semaphore.fold(zio)(_.withPermit(zio))
+
+  private def acquireConnection(using Trace): Task[Connection] =
+    ZIO
+      .attempt(dataSource.getConnection())
+      .mapError(t => SqlException("Unable to acquire DB Connection", t))
+
+  private def releaseConnection[R](con: Connection)(using Trace): URIO[R, Any] =
+    ZIO
+      .attempt(con.close())
+      .orDieWith(t =>
+        SqlException("Unable to close DB Connection, will die", t)
+      )
+end Transactor
+
+object Transactor:
+
+  /** Construct a Transactor
+    *
+    * @param dataSource
+    *   Datasource to be used
+    * @param sqlLogger
+    *   Logging configuration
+    * @param connectionConfig
+    *   Customize the underlying JDBC Connections
+    * @param maxBlockingThreads
+    *   Number of threads in your connection pool. This helps magzio be more
+    *   memory efficient by limiting the number of blocking pool threads used.
+    *   Not needed if using the ZIO virtual-thread based blocking executor
+    * @return
+    *   Transactor UIO
+    */
+  def apply(
+      dataSource: DataSource,
+      sqlLogger: SqlLogger,
+      connectionConfig: Connection => Unit,
+      maxBlockingThreads: Option[Int]
+  ): UIO[Transactor] =
```

Comment:
I think it's be more "ZIO idiomatic" if we were returning a `Layer` here.

Made a PR on top of yours to show you what it could be: https://github.com/AugustNagro/magnum/pull/88

Reply:
I agree, thanks for the code sir

Reply:
Merged it in. I did make one change to use ZLayer.fromZIO instead of Semaphore.unsafe

https://github.com/AugustNagro/magnum/pull/85/commits/7ad181a1b9388319fafc2e7e0013577cce532c58

Since I don't understand if Semaphore.unsafe is ok to use, and didn't see any examples in the docs

https://zio.dev/reference/concurrency/semaphore/

Reply:
> Since I don't understand if Semaphore.unsafe is ok to use, and didn't see any examples in the docs

It's ok to use if you know what you're doing 🙂